xrtl/gfx/util/memory_command_decoder.cc

```diff
@@ -0,0 +1,485 @@
+    switch (command_header->command_type) {
+      case CommandType::kBeginTransferCommands: {
+        reader->ReadCommand<BeginTransferCommandsCommand>(command_header);
+        transfer_encoder = target_command_buffer->BeginTransferCommands();
+        any_transfer_encoder = transfer_encoder.get();
+        any_encoder = transfer_encoder.get();
+        break;
+      }
+      case CommandType::kEndTransferCommands: {
+        reader->ReadCommand<EndTransferCommandsCommand>(command_header);
+        target_command_buffer->EndTransferCommands(std::move(transfer_encoder));
+        any_encoder = nullptr;
+        any_transfer_encoder = nullptr;
+        break;
+      }
+      case CommandType::kBeginComputeCommands: {
+        reader->ReadCommand<BeginComputeCommandsCommand>(command_header);
+        compute_encoder = target_command_buffer->BeginComputeCommands();
+        any_encoder = compute_encoder.get();
+        any_transfer_encoder = compute_encoder.get();
+        break;
+      }
+      case CommandType::kEndComputeCommands: {
+        reader->ReadCommand<EndComputeCommandsCommand>(command_header);
+        target_command_buffer->EndComputeCommands(std::move(compute_encoder));
+        any_encoder = nullptr;
+        any_transfer_encoder = nullptr;
+        break;
+      }
+      case CommandType::kBeginRenderCommands: {
+        reader->ReadCommand<BeginRenderCommandsCommand>(command_header);
+        render_encoder = target_command_buffer->BeginRenderCommands();
+        any_encoder = render_encoder.get();
+        any_transfer_encoder = render_encoder.get();
+        break;
+      }
+      case CommandType::kEndRenderCommands: {
+        reader->ReadCommand<EndRenderCommandsCommand>(command_header);
+        target_command_buffer->EndRenderCommands(std::move(render_encoder));
+        any_encoder = nullptr;
+        any_transfer_encoder = nullptr;
+        break;
+      }
+      case CommandType::kBeginRenderPass: {
+        auto command =
+            reader->ReadCommand<BeginRenderPassCommand>(command_header);
+        render_pass_encoder = target_command_buffer->BeginRenderPass(
+            ref_ptr<RenderPass>(command.render_pass),
+            ref_ptr<Framebuffer>(command.framebuffer));
+        any_encoder = render_encoder.get();
+        any_transfer_encoder = render_encoder.get();
+        break;
+      }
+      case CommandType::kEndRenderPass: {
+        reader->ReadCommand<EndRenderPassCommand>(command_header);
+        target_command_buffer->EndRenderPass(std::move(render_pass_encoder));
+        any_encoder = nullptr;
+        any_transfer_encoder = nullptr;
+        break;
+      }
+
+      case CommandType::kSetFence: {
+        auto command = reader->ReadCommand<SetFenceCommand>(command_header);
+        if (compute_encoder) {
+          compute_encoder->SetFence(ref_ptr<CommandFence>(command.fence),
+                                    command.pipeline_stage_mask);
+        } else {
```

Comment:
Could change these "else" clauses to "else if" and have a check fail in the else since it looks like fences can only be used when a compute or render encoder is set (but only one can be set at once?).
kWaitFences checks for render_pass_encoder but kSetFence and kResetFence don't?

If it's possible for user code to configure a CommandBuffer that reaches these else clauses in a bad state (hopefully that isn't the case), could bubble an error out instead of check failing.